src/analysis/retail/priest/shadow/modules/tier/Tier31ShadowPriest4Set.tsx

```diff
@@ -0,0 +1,164 @@
+import { calculateEffectiveDamage } from 'parser/core/EventCalculateLib';
+import ItemDamageDone from 'parser/ui/ItemDamageDone';
+
+const MAX_DURATION = 60000; //max duration of the buff
+const bonusSWP = 2.25; // multiplier per stack for Shadow Word Pain
+const bonusSC = 0.5; // multiplier per stack for Shadow Crash
+
+class Tier31ShadowPriest4Set extends Analyzer {
+  static dependencies = {
+    abilities: Abilities,
+    eventHistory: EventHistory,
+    spellUsable: SpellUsable,
+  };
+  protected abilities!: Abilities;
+  protected eventHistory!: EventHistory;
+  protected spellUsable!: SpellUsable;
+  has4Piece: boolean = true;
+
+  stacks = 0; //current stacks of buff
+  recentStacks = 0; // stacks used to calculate damage, since stacks are sometimes removed before damage event
+  totalStacks = 0; //total Possible Stacks
+
+  damageSC = 0; //damage to Shadow Crash
+  stacksSC = 0; //Stacks spent on Shadow Crash
+
+  stacksSWP = 0; //Stacks spent on Shadow Word Pain
+  damageSWP = 0; //damage to Shadow Word Pain
+
+  recentApplied = false; //the order of the events requires this for edge cases.
+  timestampOfLast = 0; //Used to calculate unused buffs
+
+  constructor(options: Options) {
+    super(options);
+    this.has4Piece = this.selectedCombatant.has4PieceByTier(TIERS.T31);
+    this.active = this.has4Piece;
+
+    this.addEventListener(
+      Events.applybuff.by(SELECTED_PLAYER).spell(SPELLS.SHADOW_PRIEST_TIER_31_4_SET_BUFF),
+      this.onBuffApplied,
+    );
+
+    this.addEventListener(
+      Events.applybuffstack.by(SELECTED_PLAYER).spell(SPELLS.SHADOW_PRIEST_TIER_31_4_SET_BUFF),
+      this.onBuffStack,
+    );
+
+    this.addEventListener(
+      Events.refreshbuff.by(SELECTED_PLAYER).spell(SPELLS.SHADOW_PRIEST_TIER_31_4_SET_BUFF),
+      this.onBuffRefresh,
+    );
+
+    this.addEventListener(
+      Events.cast.by(SELECTED_PLAYER).spell(TALENTS.SHADOW_CRASH_TALENT),
+      this.onShadowCrash,
+    );
+
+    this.addEventListener(
+      Events.cast.by(SELECTED_PLAYER).spell(SPELLS.SHADOW_WORD_PAIN),
+      this.onShadowWordPain,
+    );
+
+    this.addEventListener(
+      Events.damage.by(SELECTED_PLAYER).spell(TALENTS.SHADOW_WORD_DEATH_TALENT),
+      this.onShadowWordDeath,
+    );
+
+    this.addEventListener(
+      Events.removebuff.by(SELECTED_PLAYER).spell(SPELLS.SHADOW_PRIEST_TIER_31_4_SET_BUFF),
+      this.onBuffRemoved,
+    );
+  }
+
+  onBuffApplied(Event: ApplyBuffEvent) {
+    this.timestampOfLast = Event.timestamp;
+    this.stacks = 1;
+    this.recentApplied = true;
+  }
+
+  onBuffStack(Event: ApplyBuffStackEvent) {
+    this.timestampOfLast = Event.timestamp;
+    this.stacks = Event.stack;
+  }
+
+  onBuffRefresh(Event: RefreshBuffEvent) {
+    this.timestampOfLast = Event.timestamp;
+  }
+
+  onBuffRemoved(Event: RemoveBuffEvent) {
+    this.recentStacks = this.stacks;
+    if (MAX_DURATION <= Event.timestamp - this.timestampOfLast) {
+      this.recentStacks = 0;
+    }
+    this.stacks = 0;
+  }
```

Comment:
Unless I'm missing something, all this is superfluous now and can be removed

Reply:
I think I still need to calculate the number of stacks using `OnBuffApplied `and `OnBuffStack`, but otherwise you are right.  I thought I needed to make sure I had the right number of stacks in edge cases, but the normalizer handles all of them.

Reply:
Oh yeah you do still need the stack count. The 'recent applied' stuff is handled by the normalizer because we do the tally at the timestamp the cast happens instead of waiting for the damage.